### PR TITLE
i#5975: Add invariant check that read/write records match operands

### LINKED ISSUE #5975: Add drmemtrace invariant checks that read/write records match operands
Now that we have the decoder linked into the invariant checker we can add more invariant checks: making sure the count and type of read and write records following each instruction matches its operands.

### PR DESCRIPTION
**Pull request by ivankyluk**
Add an invariant check to verify the number of memory read/write records is matching the operands.

Update check_sane_control_flow() to use  instrlist_t and memref_with_IR_t to avoid hardcoded encodings. Current tests fail the new check since they have memory accesses.

Fixes #5975

### CODE REVIEW

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 1]
```diff
@@ -281,13 +281,15 @@ check_sane_control_flow()
             gen_marker(TID, TRACE_MARKER_TYPE_FILETYPE, OFFLINE_FILE_TYPE_ENCODINGS),
 #    if defined(X86_64) || defined(X86_32)
             // 0x74 is "je" with the 2nd byte the offset.
-            gen_branch_encoded(TID, 0x71019dbc, { 0x74, 0x32 }),
-            gen_instr_encoded(0x71019ded, { 0x01 }, TID),
+            gen_branch_encoded(TID, /*pc=*/0x71019dbc, /*encoding=*/ { 0x74, 0x32 }),
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp:284]: The statement

gen_instr_encoded(0x71019ded, { 0x01 }, TID)

causes an expectation of one read and one write memory access. Since 0x01 is invalid, I changed it to nop to proceed.

I can investigate why it's the case. Since it happens only in test, I'd consider it low priority. Please let me know if you think otherwise.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 2]
```diff
@@ -969,14 +1034,15 @@ invariant_checker_t::check_for_pc_discontinuity(
...
             // TODO i#5912: This prev_instr_decoded_ isn't always the same as the
-            // decoding of the passed-in prev_instr, for signals.  What we should do is
-            // create a new instr_info_t struct with fields for the memref_t plus the
-            // from-decoding attributes is_syscall, writes_memory, and branch_target, and
-            // fill it in at decode time.  We can then store it as prev_instr_ and in
-            // pre_signal_instr and last_instr_in_cur_context_ and pass it here as
-            // prev_instr.  We can then delete this branch_target_cache and
-            // prev_instr_decoded_.  For now we live with this inaccuracy as it is
-            // rare to not match and we leave the cleanup for i#5912.
+            // decoding of the passed-in prev_instr, for signals.  What we should do
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply clients/drcachesim/tools/invariant_checker.cpp:1037]: I'll update the clang-format rules.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/invariant_checker_test.cpp:284]: OK, I see.  I think I didn't understand the PR description mentioning this: maybe reword that?

[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/invariant_checker_test.cpp:284]: Eventually we'd like to replace these with an instrlist too; could be done now.

File: clients/drcachesim/common/trace_entry.h [hunk 3]
```diff
@@ -634,6 +634,15 @@ type_is_data(const trace_type_t type)
         type == TRACE_TYPE_DATA_FLUSH_END;
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply clients/drcachesim/common/trace_entry.h:634]: Done.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp]: I've updated the code to use instrlist here.

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 4]
```diff
@@ -1644,6 +1649,231 @@ check_timestamps_increase_monotonically(void)
...
+    // Correct: number of read records matches operand.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_marker(TID, TRACE_MARKER_TYPE_FILETYPE, OFFLINE_FILE_TYPE_ENCODINGS),
+#if defined(X86_64) || defined(X86_32)
+            // 0x7fcf3b9d: 83 3d 84 21 23 03 00 cmp    <rel> 0x000000000443ddc0[4byte]
+            // $0x00000000
+            gen_instr_encoded(ADDR,
+                              /*encoding=*/
+                              { 0x83, 0x3d, 0x84, 0x21, 0x23, 0x03, 0x00 }),
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp]: Done. PTAL.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 5]
```diff
@@ -519,8 +553,9 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
         // Invariant: offline traces guarantee that a branch target must immediately
         // follow the branch w/ no intervening thread switch.
         // If we did serial analyses only, we'd just track the previous instr in the
-        // interleaved stream.  Here we look for headers indicating where an interleaved
-        // stream *could* switch threads, so we're stricter than necessary.
+        // interleaved stream.  Here we look for headers indicating where an
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply clients/drcachesim/tools/invariant_checker.cpp]: I ran clang-format in a sub directory, hence the issue. I've reverted the changes.

File: core/ir/x86/decode_table.c [hunk 6]
```diff
@@ -9533,97 +9533,97 @@ const instr_info_t rep_extensions[][4] = {
...
     {OP_cmps,       0xa70000, "cmps",         axSI, axDI, Xv, Yv, axSI, xop_next, (fW6|fRD), tne[0][0]},
     {OP_CONTD,      0xa70000, "cmps",         xx, xx, axDI, xx, xx, no, (fW6|fRD), END_LIST},
-    {OP_rep_cmps,   0xf3a70000, "rep cmps",   axSI, axDI, Xv, Yv, axSI, xop_next, (fW6|fRD|fRZ), tne[0][2]},
-    {OP_CONTD,      0xf3a70000, "rep cmps",   axCX, xx, axDI, axCX, xx, no, (fW6|fRD), END_LIST},
-    {OP_repne_cmps, 0xf2a70000, "repne cmps", axSI, axDI, Xv, Yv, axSI, xop_next, (fW6|fRD|fRZ), tne[0][4]},
-    {OP_CONTD,      0xf2a70000, "repne cmps", axCX, xx, axDI, axCX, xx, no, (fW6|fRD), END_LIST},
+    {OP_rep_cmps,   0xf3a70000, "rep cmps",   axSI, axDI, Xv, Yv, axSI, xop_next|predcx, (fW6|fRD|fRZ), tne[0][2]},
+    {OP_CONTD,      0xf3a70000, "rep cmps",   axCX, xx, axDI, axCX, xx, predcx, (fW6|fRD), END_LIST},
+    {OP_repne_cmps, 0xf2a70000, "repne cmps", axSI, axDI, Xv, Yv, axSI, xop_next|predcx, (fW6|fRD|fRZ), tne[0][4]},
+    {OP_CONTD,      0xf2a70000, "repne cmps", axCX, xx, axDI, axCX, xx, predcx, (fW6|fRD), END_LIST},
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply core/ir/x86/decode_table.c]: Changes related to marking x86 rep as predicated have been moved, and submitted through PR6284 (https://github.com/DynamoRIO/dynamorio/pull/6284).

File: core/ir/x86/decode_table.c [hunk 7]
```diff
@@ -9533,97 +9533,97 @@ const instr_info_t rep_extensions[][4] = {
   { /* rep extension 0 */
     {OP_ins,      0x6c0000, "ins",       Yb, axDI, dx, axDI, xx, no, fRD, END_LIST},
     {INVALID,   0x00000000, "(bad)",  xx, xx, xx, xx, xx, no, x, NA},
-    {OP_rep_ins,  0xf36c0000, "rep ins", Yb, axDI, dx, axDI, axCX, xop_next, fRD, END_LIST},
-    {OP_CONTD,  0xf36c0000, "rep ins", axCX, xx, xx, xx, xx, no, fRD, END_LIST},
+    {OP_rep_ins,  0xf36c0000, "rep ins", Yb, axDI, dx, axDI, axCX, xop_next|predcx, fRD, END_LIST},
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply core/ir/x86/decode_table.c]: Changes related to marking x86 rep as predicated have been moved, and submitted through PR6284 (https://github.com/DynamoRIO/dynamorio/pull/6284).

[the remark this answers is not in the document]
Comment (ivankyluk) [reply core/ir/x86/decode_table.c]: Changes related to marking x86 rep as predicated have been moved, and submitted through PR6284 (https://github.com/DynamoRIO/dynamorio/pull/6284).

File: core/ir/instr_shared.c [hunk 8]
```diff
@@ -52,12 +52,14 @@
 #define INSTR_INLINE extern inline
 
 #include "../globals.h"
-#include "instr.h"
+
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply core/ir/instr_shared.c]: I've reverted the include file changes.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ivankyluk) [reply clients/drcachesim/tools/invariant_checker.cpp]: I've reverted the format change.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 9]
```diff
@@ -811,6 +848,34 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
     shard->prev_entry_ = memref;
     if (type_is_instr_branch(shard->prev_entry_.instr.type))
         shard->last_branch_ = shard->prev_entry_;
+
+    if (type_is_data(memref.data.type) && shard->prev_instr_decoded_ != nullptr) {
+        if (!instr_is_predicated(shard->prev_instr_decoded_->data)) {
+            if (type_is_read(memref.data.type)) {
+                // Skip D-filtered traces which don't have every load or store records.
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply clients/drcachesim/tools/invariant_checker.cpp:852]: done.

File: core/ir/instr_api.h [hunk 10]
```diff
@@ -1656,6 +1656,20 @@ DR_API
 uint
 instr_memory_reference_size(instr_t *instr);
 
+DR_API
+/**
+ * Returns the number of memory read accesses of the instruction.
+ */
+uint
+instr_num_memory_read_access(instr_t *instr);
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply core/ir/instr_api.h:1664]: done.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 11]
```diff
@@ -811,6 +848,34 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
     shard->prev_entry_ = memref;
     if (type_is_instr_branch(shard->prev_entry_.instr.type))
         shard->last_branch_ = shard->prev_entry_;
+
+    if (type_is_data(memref.data.type) && shard->prev_instr_decoded_ != nullptr) {
+        if (!instr_is_predicated(shard->prev_instr_decoded_->data)) {
```
[the remark this answers is not in the document]
Comment (ivankyluk) [reply clients/drcachesim/tools/invariant_checker.cpp]: done.

File: api/docs/release.dox [hunk 12]
```diff
@@ -139,6 +139,9 @@ Further non-compatibility-affecting changes include:
    -core_serial) and control the schedule
```
Comment (derekbruening) [inline api/docs/release.dox:139]: > …cords matching operands.

As you can see, the title is truncated and wrapped around, which is not ideal.  Please tighten up the title to avoid truncation.  Maybe "i#5975: Add invariant check that read/write records match operands".

File: clients/drcachesim/common/trace_entry.h [hunk 13]
```diff
@@ -634,6 +634,15 @@ type_is_data(const trace_type_t type)
...
 }
 
+/**
+ * Returns whether the type represents a memory read access.
+ */
+static inline bool
+type_is_read(const trace_type_t type)
+{
+    return type_is_prefetch(type) || type == TRACE_TYPE_READ;
```
Comment (derekbruening) [inline clients/drcachesim/common/trace_entry.h:643]: Hmm, what about cache flushes which are part of instr_is_data() and instr_has_address()?  If you leave them out of here, what if someone uses instr_is_data() or instr_has_address() and then assumes !type_is_read() means it's a store and so assumes all the cache flushes are stores?  If left out, probably the doc comment should explicitly say it excludes them, to avoid confusion vs type_has_address().

UPDATE: I see your own code does this!

As for whether we should check cache flushes: certainly x86 OP_clflush has a memory operand (marked as a source, so a load/read) we can check.  There are unresolved issues with the a64 decoder though IIRC: #4400 has discussion and linked issues/PRs.  It might be messy to try to check them if the deoder is inconsistent -- but OTOH if the instrumentation is using the same operands to get the address for the record it should work in the checker too.

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 14]
```diff
@@ -1872,6 +1882,201 @@ check_timestamps_increase_monotonically(void)
...
+              nullptr },
+            { gen_instr(TID), lea },
+            { gen_instr(TID), nop },
+        };
+        auto memrefs = add_encodings_to_memrefs(ilist, memref_setup, BASE_ADDR);
+        instrlist_clear_and_destroy(GLOBAL_DCONTEXT, ilist);
+        if (!run_checker(memrefs, false)) {
+            return false;
+        }
+    }
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:2075]: How about a test of an instruction with both read and write operands (e.g., x86 push from memory or movs).

Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:2075]: Please add at least one test of cache flush instructions (see type_is_read() comment).

**Review by derekbruening — APPROVED** [review]
Cache flushes are not handled properly but probably doesn't need another review.

File: api/docs/release.dox (hunk 12 above)
Comment (ivankyluk) [reply api/docs/release.dox]: done

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 14 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp:2199]: done.

Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp:2199]: done.

File: clients/drcachesim/common/trace_entry.h (hunk 13 above)
Comment (ivankyluk) [reply clients/drcachesim/common/trace_entry.h]: I've added cache flushes to type_is_read.

File: api/docs/release.dox [hunk 15]
```diff
@@ -141,6 +141,9 @@ Further non-compatibility-affecting changes include:
    -replay_file, -cpu_schedule_file).
```
Comment (derekbruening) [inline api/docs/release.dox:141]: > …cords matching operands.

Noticed that this is still in the description.

Comment (ivankyluk) [reply api/docs/release.dox:141]: Thanks for catching it.
